cns/service.go

```diff
@@ -48,64 +47,93 @@ func NewService(name, version, channelMode string, store store.KeyValueStore) (*
 	}, nil
 }
 
-// GetAPIServerURL returns the API server URL.
-func (service *Service) getAPIServerURL() string {
-	urls, _ := service.GetOption(acn.OptCnsURL).(string)
-	if urls == "" {
-		urls = defaultAPIServerURL
-	}
-
-	return urls
-}
-
-// Initialize initializes the service and starts the listener.
-func (service *Service) Initialize(config *common.ServiceConfig) error {
-	log.Debugf("[Azure CNS] Going to initialize a service with config: %+v", config)
-
-	// Initialize the base service.
-	if err := service.Service.Initialize(config); err != nil {
-		return errors.Wrap(err, "failed to initialize")
-	}
+func (service *Service) AddListener(config *common.ServiceConfig) error {
+	var (
+		err     error
+		nodeURL *url.URL
+	)
+
+	// if cnsURL is empty the VM primary interface IP will be used
+	// if customer specifies -c option, then use this URL with warning message and it will be deprecated soon
+	cnsURL, _ := service.GetOption(acn.OptCnsURL).(string)
+	// if customer provides port number by -p option, then use VM IP with this port and localhost server also uses this port
+	// otherwise it will use defaultAPIServerPort 10090
+	cnsPort, _ := service.GetOption(acn.OptCnsPort).(string)
+
+	if cnsURL == "" {
+		config.Server.EnableLocalServer = true
```

Comment:
I generally think it's bad practice to write back to a config received from the environment--that should be strictly immutable.

Reply:
We will get rid of this flag once -c option is deprecated